Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,265 @@
+/// # Parser combinators in Dafny
+///
+/// Challenge: define parse combinators.  As a usability test, use them to parse
+/// expressions of the form `(ⁿ)ⁿ` (`n` opening parentheses followed by `n`
+/// closing parentheses).
+///
+/// This challenge is tricky because we can't write what we'd usually write in
+/// Haskell or OCaml:
+///
+/// ```ocaml
+/// let parentheses' () =
+///   (either
+///      (concat (char '(') (concat parentheses' (char ')')))
+///      epsilon)
+/// let parentheses () =
+///   (concat parentheses' eos)
+/// ```
+///
+/// Dafny will reject this definintion definition because thunking isn't enough
```

Comment:
`definition definition`?

Reply:
Also: "thunking" as a verb might be a bit arcane, "pushing the recursion under a function definition" maybe?